engine/common/follower/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+	delete(c.byView, block.Header.View)
+	delete(c.byParent, block.Header.ParentID)
+}
+
+// AddBlocks atomically applies batch of blocks to the cache of pending but not yet certified blocks. Upon insertion cache tries to resolve
+// incoming blocks to what is stored in the cache.
+// We require that incoming batch is sorted by height and doesn't have skipped blocks.
+// When receiving batch: [first, ..., last], we are only interested in first and last blocks since all other blocks will be certified by definition.
+// Next scenarios are possible:
+// - for first block:
+//   - no parent available for first block, we need to cache it since it will be used to certify parent when it's available.
+//   - parent for first block available in cache allowing to certify it, no need to store first block in cache.
+//
+// - for last block:
+//   - no child available for last block, we need to cache it since it's not certified yet.
+//   - child for last block available in cache allowing to certify it, no need to store last block in cache.
+//
+// Note that implementation behaves correctly where len(batch) == 1.
+// If message equivocation was detected it will be reported using a notification.
+// Concurrency safe.
+func (c *Cache) AddBlocks(batch []*flow.Block) (certifiedBatch []*flow.Block, certifyingQC *flow.QuorumCertificate) {
+	var equivocatedBlocks [][]*flow.Block
+
+	// prefill certifiedBatch with minimum viable result
+	// since batch is a chain of blocks, then by definition all except the last one
+	// has to be certified by definition
+	certifiedBatch = batch[:len(batch)-1]
+
+	if len(batch) > 1 {
+		// set certifyingQC, QC from last block certifies complete batch
+		certifyingQC = batch[len(batch)-1].Header.QuorumCertificate()
+	}
+
+	c.lock.Lock()
+	// check for message equivocation, report any if detected
+	for _, block := range batch {
+		if otherBlock, ok := c.byView[block.Header.View]; ok {
+			if otherBlock.ID() != block.ID() {
+				equivocatedBlocks = append(equivocatedBlocks, []*flow.Block{otherBlock, block})
+			}
+		} else {
+			c.byView[block.Header.View] = block
+		}
+		// store all blocks in the cache to provide deduplication
+		c.backend.Add(block.ID(), block)
+		c.byParent[block.Header.ParentID] = block
+	}
+
+	firstBlock := batch[0]           // lowest height/view
+	lastBlock := batch[len(batch)-1] // highest height/view
+
+	// start by checking if batch certifies any block that was stored in the cache
+	if parent, ok := c.backend.ByID(firstBlock.Header.ParentID); ok {
+		// parent found, it can be certified by the batch, we need to include it to the certified blocks
+		certifiedBatch = append([]*flow.Block{parent.(*flow.Block)}, certifiedBatch...)
+		// set certifyingQC, QC from last block certifies complete batch
+		certifyingQC = batch[len(batch)-1].Header.QuorumCertificate()
+	}
+
+	// check if there is a block in cache that certifies last block of the batch.
+	if child, ok := c.byParent[lastBlock.ID()]; ok {
```

Comment:
I've moved it outside of critical section, I think in long run the correct solution is to avoid computing `ID` on each call, when we will change how blocks are accepted we will be able to achieve this optimization seamlessly